runtime/lua/vim/filetype/detect.lua

```diff
@@ -787,9 +862,67 @@ function M.scd(bufnr)
   end
 end
 
-function M.sh(path, bufnr) end
+-- Also called from filetype.lua
+function M.sh(path, bufnr, name)
+  if vim.fn.did_filetype() or path:find(vim.g.ft_ignore_pat) then
+    -- Filetype was already detected or detection should be skipped
+    return
+  end
 
-function M.shell(path, bufnr) end
+  if matchregex(name, [[\c\<csh\>]]) then
+    -- Some .sh scripts contain #!/bin/csh.
+    M.shell(path, bufnr, 'csh')
+    return
+    -- Some .sh scripts contain #!/bin/tcsh.
+  elseif matchregex(name, [[\c\<tcsh\>]]) then
+    M.shell(path, bufnr, 'tcsh')
+    return
+    -- Some .sh scripts contain #!/bin/zsh.
+  elseif matchregex(name, [[\c\<zsh\>]]) then
+    M.shell(path, bufnr, 'zsh')
+    return
+  elseif matchregex(name, [[\c\<ksh\>]]) then
+    vim.b[bufnr].is_kornshell = 1
+    vim.b[bufnr].is_bash = nil
+    vim.b[bufnr].is_sh = nil
+  elseif
+    (vim.g.bash_is_sh and vim.g.bash_is_sh ~= 0)
```

Comment:
filetype.vim only checks for existence, not the value

```suggestion
    vim.g.bash_is_sh
```